src/BizHawk.Client.Common/movie/import/IMovieImport.cs

```diff
@@ -60,12 +52,65 @@ protected void MaybeSetCorePreference(string sysID, string coreName, string file
 
 			if (!Result.Errors.Any())
 			{
+				if (string.IsNullOrEmpty(Result.Movie.Hash))
+				{
+					string hash = null;
+					// try to generate a matching hash from the original ROM
+					if (Result.Movie.HeaderEntries.TryGetValue("CRC32", out string crcHash))
+					{
+						hash = PromptForRom(data => string.Equals(CRC32Checksum.ComputeDigestHex(data), crcHash, StringComparison.OrdinalIgnoreCase));
+					}
+					else if (Result.Movie.HeaderEntries.TryGetValue("MD5", out string md5Hash))
+					{
+						hash = PromptForRom(data => string.Equals(MD5Checksum.ComputeDigestHex(data), md5Hash, StringComparison.OrdinalIgnoreCase));
+					}
+					else if (Result.Movie.HeaderEntries.TryGetValue("SHA256", out string sha256Hash))
+					{
+						hash = PromptForRom(data => string.Equals(SHA256Checksum.ComputeDigestHex(data), sha256Hash, StringComparison.OrdinalIgnoreCase));
+					}
+
+					if (hash is not null)
+						Result.Movie.Hash = hash;
+				}
+
 				Result.Movie.Save();
 			}
 
 			return Result;
 		}
 
+		/// <summary>
+		/// Prompts the user for a ROM file that matches the original movie file's hash
+		/// and returns a SHA1 hash of that ROM file.
+		/// </summary>
+		/// <param name="matchesMovieHash">Function that checks whether the ROM data matches the original hash</param>
+		/// <returns>SHA1 hash of the selected ROM file</returns>
+		private string PromptForRom(MatchesMovieHash matchesMovieHash)
+		{
+			string messageBoxText = "Please select the original ROM to finalize the import process.";
+			while (true)
+			{
+				if (!_dialogParent.ModalMessageBox2(messageBoxText, "ROM required to populate hash", useOKCancel: true))
+					return null;
+
+				var result = _dialogParent.ShowFileOpenDialog(
+					filter: RomLoader.RomFilter,
+					initDir: Config.PathEntries.RomAbsolutePath(_emulator.SystemId));
+				if (result is null)
+					return null;
+
+				using var rom = new HawkFile(result);
+				if (rom.IsArchive) rom.BindFirst();
+				ReadOnlySpan<byte> romData = rom.ReadAllBytes();

```

Comment:
Outside method calls, prefer `AsSpan` or w/e it's called to implicit cast. I presume you only did it this way because you're conditionally mutating the local on the next line.

Reply:
What's the advantage of `AsSpan()` over the implicit cast? Is it just for readability purposes?

The delegate takes a `ReadOnlySpan` also, so doing `.AsSpan()` would create a normal `Span` and then later implicitly cast this to a `ReadOnlySpan`.

Reply:
> Is it just for readability purposes?

Yes